### PR TITLE
[Merged by Bors] - feat: add support for cluster metadata

### LINKED ISSUE #3627: Allow annotations in fluvio profiles
The Fluvio profile should have an annotations section which allows other tools to "tag" or "label". I imagine this should work just like kubernetes object annotations.

Example 1:
```
[profile.<label>.annotations]
fluvio.io/cluster-type = "local-k8"
```

Example 2:
```
[profile.<label>.annotations]
infinyon.cloud/environment = "infinyon.cloud"
infinyon.cloud/cluster-name = "my-cloud-cluster-name"
```

### PR DESCRIPTION
**Pull request by matheus-consoli**
Add support for custom metadata in the cluster configuration.

```toml
[cluster.my_favorite_cluster]
endpoint = "127.0.0.1:9003"

[cluster.my_favorite_cluster.metadata.installation]
"fluvio.io/cluster-type" = "local-k8"
```

This will be important for other tasks, such as #3620.

These changes are retrocompatible.

Closes #3627

### CODE REVIEW

**Comment by sehz** [comment]
This just adds more noise.  

Instead add cluster type to config object so can avoid parsing error

**Comment by sehz** [comment]
[cluster.my_favorite_cluster]
endpoint = "127.0.0.1:9003"
type = "local-k8"

**Comment by matheus-consoli** [comment]
> Instead add cluster type to config object so can avoid parsing error

That would work for the cluster-type issue, but I think @galibey also has a use case for this same feature (cmiiw @galibey)

**Comment by sehz** [comment]
then lets' add more properties as necessary.  annotations doesn't buy anything except adding more friction and increase complexity

**Comment by sehz** [comment]
if there are more properties to add, then it probably make sense to move profile as separate crate to make it more reusable

**Comment by digikata** [comment]
The eng team already discussed this in the engineering meeting and prefers to use annotations.  There is a tradeoff between hard coded properties and annotations. One benefit of annotation is that there is less coupling between different versions later. I think as long as Nick, Alex, and Matheus agree on the implementation detail, we can let them move forward.

**Review by nacardin — COMMENTED** [review]
Why not add this under the `Profile` struct instead?  It would be easier to match to a profile if it was at the `Profile` level. Putting it at the `FluvioConfig` level requires querying the cluster reference while iterating through profiles.

File: crates/fluvio/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "fluvio"
-version = "0.21.1"
+version = "0.21.2"
```
Comment (sehz) [inline crates/fluvio/Cargo.toml:3]: 21.1 is not release, so no need to bump

**Review by sehz — CHANGES_REQUESTED** [review]
Overall LGTM.  minor nit

**Comment by galibey** [comment]
What is the point of having `toml::Value` field? It is not so much better than just vec of bytes, it is too loose type in my opinion. Plus, we couple our serialization format (toml) to public API.
If we choose to no go with strictly type fields in the cluster config, we could use some `MetadataAnnotation` enum that encapsulates different possible representations there and we expose `metadata: Vec<MetadataAnnotation>` instead of `pub metadata: Option<toml::Value>`

**Comment by sehz** [comment]
This is similar to the approach cargo metadata: https://docs.rs/cargo_metadata/0.18.1/cargo_metadata/struct.Metadata.html.   The whole point of annotation is to add opinionated config to config without adding more dep to the Fluvio crate.  It's the same problem, even if this moves to a separate crate.   Then, it is the responsibility of the reader/producer of annotation to ensure the config is correct.  With serde `Value`,  they can choose to use strong type or not.

File: crates/fluvio/Cargo.toml (hunk 1 above)
Comment (matheus-consoli) [reply crates/fluvio/Cargo.toml:3]: check crate-version [failed](https://github.com/infinyon/fluvio/actions/runs/6702295994/job/18211081175?pr=3628), I had to change it back to 0.21.2

**Review by sehz — APPROVED** [review]
LGTM

**Comment by matheus-consoli** [comment]
the [first attempt failed](https://github.com/infinyon/fluvio/actions/runs/6712710226/job/18243007947) because of a socket timeout, trying again

bors r+

**Comment by galibey** [comment]
@sehz Imagine you need access to some metadata item from 3 different crates: you have to have `serde_toml` in all these three places, and you have to do deserialization there, also, you need that target type dependency (or use code duplication). So, it will be something like `let my_type: MyType = serde_toml::from_value(config.metadata)` everywhere. It is not convenient at all.

**Comment by matheus-consoli** [comment]
@galibey I think this is a valid concern, what do you think if we add a way for querying metadata?

I think it would look something like `pub fn query_metadata<'de, T: Deserialize<'de>>(&self, query: &str) -> Option<T>`, with a micro-dls to target specific fields from some metadata table.

for example, consider the table below and a query to get `"value"`:

```toml
[cluster.name.metadata.deep.nested]
key = "value"
```
```rust
assert_eq!(cluster.query_metadata::<&str>("deep.nested[key]"), Some("value"));
```

If we agree on this, it will target all your concerns to some extent:

- [x] no need to add `toml` as a dependency everywhere
- [x] (in some cases) we don't need the target type or code duplication
  - we can query specific fields, or deserialize them into some `map`-structure
- [x] is more convinient

some downsides:
- cost of search
- stringly-typed

**Comment by galibey** [comment]
@matheus-consoli Yes, that would work. I like the idea.

**Comment by galibey** [comment]
We will also need a method for inserting a metadata entry.

**Comment by sehz** [comment]
Let's not add more complexity to this API and more work.   If some client wants to add strong typing, they can create a wrapper around it which at that point, can hide all dependency.

I agree with @galibey, inserting and updating is needed.

File: crates/fluvio/src/config/cluster.rs [hunk 2]
```diff
@@ -58,6 +62,74 @@ impl FluvioConfig {
         self.tls = tls.into();
         self
     }
+
+    pub fn query_metadata_path<'de, T>(&self, path: &str) -> Option<T>
+    where
+        T: Deserialize<'de>,
+    {
+        let mut metadata = self.metadata.as_ref().unwrap();
```
Comment (galibey) [inline crates/fluvio/src/config/cluster.rs:70]: This is not good. If there is no metadata, the querying here will shut my process.

Comment (matheus-consoli) [reply crates/fluvio/src/config/cluster.rs:70]: ohno, you're right, I missed that one!

**Pull request closed without merging**